core/trino-main/src/main/java/io/trino/execution/scheduler/PipelinedStageExecution.java

```diff
@@ -366,25 +390,72 @@ private synchronized void updateTaskStatus(TaskStatus taskStatus)
                 // A task should only be in the aborted state if the STAGE is done (ABORTED or FAILED)
                 fail(new TrinoException(GENERIC_INTERNAL_ERROR, "A task is in the ABORTED state but stage is " + stageState));
                 break;
-            case FLUSHING:
-                flushingTasks.add(taskStatus.getTaskId());
-                break;
-            case FINISHED:
-                finishedTasks.add(taskStatus.getTaskId());
-                flushingTasks.remove(taskStatus.getTaskId());
-                break;
             default:
+                break;
+        }
+    }
+
+    private void updateTaskStatus(TaskStatus taskStatus)
+    {
+        State stageState = stateMachine.getState();
+        TaskState taskState = taskStatus.getState();
+        if (stageState.isDone()) {
+            return;
+        }
+
+        if (taskState == TaskState.FAILED || taskState == TaskState.CANCELED || taskState == TaskState.ABORTED) {
+            updateNotSuccessfulTaskStatus(taskStatus);
+            return;
+        }
+
+        long flushingOrFinishedTasksCounter = 0;
+        long finishedTasksCounter = 0;
+
+        /*
+         * We are counting the unique number of flushing/finished tasks using the atomic counter flushingOrFinishedTasksCounter.
+         * When some thread detects that number of finished or flushing tasks is equal
+         * to allTasks.size() then it decides to move stateMachine's state to FLUSHING/FINISHED. To distinguish between
+         * FLUSHING or FINISHED we use `flushingTaskWasObserved` to mark that there was at least one flushing task.
+        */
+        if (taskState == TaskState.FLUSHING) {
+            boolean addedNow = flushingOrFinishedTasks.add(taskStatus.getTaskId());
+            if (addedNow) {
+                flushingTaskWasObserved = true;
+                flushingOrFinishedTasksCounter = this.flushingOrFinishedTasksCounter.incrementAndGet();
+            }
+        }
+
+        /*
+         * We are counting the unique number of finished tasks using the atomic counter finishedTasksCounter.
+         * When some thread detects that number of finished tasks is equal to allTasks.size() then it decides
+         * to move stateMachine's state to FINISHED.
+         */
+        else if (taskState == TaskState.FINISHED) {
+            boolean addedNow = flushingOrFinishedTasks.add(taskStatus.getTaskId());
+            if (addedNow) {
+                flushingOrFinishedTasksCounter = this.flushingOrFinishedTasksCounter.incrementAndGet();
```

Comment:
you are only incrementing `flushingOrFinishedTasksCounter` here and at line 424 but never decrement. Can the task be first `FLUSHING` and then `FINISHED`? if so, you will double count it I think.

Reply:
It is not possible that we double count something because we count only when the task was put to the set the first time. 
It is possible that for `FINISHED` status it will be counted in the `flushingOrFinishedTasksCounter` and `finishedTasksCounter` but it is ok because we actually wants to count it.